### PR TITLE
Fix albumsQuery artist and artist_id

### PR DESCRIPTION
**Pull request by darrell-k**
See https://github.com/LMS-Community/slimserver/issues/1205

### CODE REVIEW

**Review by michaelherger — COMMENTED** [review]
This doesn't really fix the issue, does it? It does check for the existence of `$contributorID` and `$c->{albums.contributor}` only for the name, but not for the ID. The latter would always be the ID from the query response. The problem seems to be inverted (always ID from database, rather than query instead of the opposite), but not fixed?

 It would _always_ use the ID from the database,

**Comment by darrell-k** [comment]
albums.contributor will always be consistent as it's a column from the base albums table. The problem is that contributor.name from the query is from the contributors table (via a join with contributor_album) where there is a match on the input parm.

It certainly took me a few brain cycles to get my head around this, but my testing of 3 scenarios (no input contributor, input contributor same as album contributor and input contributor different to album contributor) makes me believe I've got this right.

But feel free to prove me wrong!

**Comment by michaelherger** [comment]
My simplistic take is: whatever was passed, the `artist_id` should match the `artist`. Under what circumstances would we have a `contributor.name` which was not matching `artist.contributor`? Wouldn't that mean there's a problem further up the code, in the initial query?

Would you have an example how I can reproduce the problem?

**Comment by darrell-k** [comment]
You need to browse an artist who appears on the album but is not the main artist (as defined by albums.contributor). A trackartist, perhaps? (I've seen it with user-defined tags, and conductor, so you could just add "Conductor=Michael" to an album.)

Then use curl to submit a jsonrpc request, because I'm still not sure how to see the bug in the UI.

You make an interesting point about the initial query SQL: should it join to contributors on albums.contributor or contributor_album.contributor for contributors.name? It might do either, depending on the input parameters, including `$sort`.

And whatever we get in contributor.name is used in generating the favoritesUrl which seems to work correctly at present, but I'll check with various parameters to trigger both cases.

(presumably when you said artist.contributor above you meant albums.contributor?)

**Comment by darrell-k** [comment]
Actually, Favorites are NOT working when added in this case - it definitely needs the name associated with albums.contributor.

Another ancient bug referenced here:
https://github.com/LMS-Community/slimserver/blob/a1719ae7ab4d8a6238e958781196931fd7c9ec9c/Slim/Control/Queries.pm#L617-L633

Maybe we could join to contributors twice, once on albums.contributor and again on contributor_album.contributor, to get both names and use as appropriate in the result loop processing to avoid that extra SQL call in the loop?

**Comment by darrell-k** [comment]
OK, I think that latest push is the solution.

One other point: should creating `favorites_url` be dependent on tag `a`, because it needs data for `contributor.name`? (This isn't a change - look at the existing code, it will try to create `favorites_url` without tag `a` already.)

**Comment by michaelherger** [comment]
Did you run any performance tests? That's creating a (potentially) second relation to the contributors table, which will come at a cost, right?

**Comment by michaelherger** [comment]
> One other point: should creating `favorites_url` be dependent on tag `a`, because it needs data for `contributor.name`? (This isn't a change - look at the existing code, it will try to create `favorites_url` without tag `a` already.)

It shouldn't depend on that tag, no. Which probably means everything can be simplified further? We'll always need that relation?

**Comment by darrell-k** [comment]
I would think performance would be better.  For tag `a`, we're adding a extra join in the main sql, instead of running an extra query for every album in the in the loop.

This remains true even if we remove the conditioning on tag `a`.

**Comment by michaelherger** [comment]
True, I should have asked the question before you added the additional query in the loop 😁.

Are you doing to look into the unconditional relation?

**Comment by darrell-k** [comment]
> True, I should have asked the question before you added the additional query in the loop 😁.
> 
> Are you doing to look into the unconditional relation?

To be fair to me, the extra query in the loop was already there 😁.

Yes, I'm going to do the new join always unless tags = `CC`.

Also, I'm looking at
https://github.com/LMS-Community/slimserver/blob/a1719ae7ab4d8a6238e958781196931fd7c9ec9c/Slim/Control/Queries.pm#L460-L473
and
https://github.com/LMS-Community/slimserver/blob/a1719ae7ab4d8a6238e958781196931fd7c9ec9c/Slim/Control/Queries.pm#L866-L877

If the sort should always be using the `albums.contributor` name here, rather than the name of the input param `contributorID` if supplied, I think we can get rid of the join to `contributors` via `contributor_album` completely, because we'd only be using `contributorID` to restrict the result set, never needing the name.

What do you think?

**Comment by michaelherger** [comment]
> To be fair to me, the extra query in the loop was already there 😁.

Oh... I thought you had introduced that one. But no, was me...

> If the sort should always be using the albums.contributor name here, rather than the name of the input param contributorID if supplied, I think we can get rid of the join to contributors via contributor_album completely, because we'd only be using contributorID to restrict the result set, never needing the name.

Except if some particular role has to be dealt with. Or if we want the full list of contributors, right?

**Comment by darrell-k** [comment]
> > If the sort should always be using the albums.contributor name here, rather than the name of the input param contributorID if supplied, I think we can get rid of the join to contributors via contributor_album completely, because we'd only be using contributorID to restrict the result set, never needing the name.
> 
> Except if some particular role has to be dealt with. Or if we want the full list of contributors, right?

We'd still join to `contributor_album` if necessary for row selection, it's just that we'd never need to get name/namesort for that.

The lists of contributors/names are retrieved separately in `$contributorSql`.

Agreed?

**Comment by darrell-k** [comment]
@michaelherger I've additional changes lined up for this, based on the last couple of messages here. Should I commit & push, or is there a flaw in my reasoning?

**Comment by darrell-k** [comment]
Ah just saw the thumbs-up in the previous message! I'll push.

**Comment by michaelherger** [comment]
Let's try to slow down and stabilise things. We shouldn't shake existing code too much any more, even if it's a more elegant solution or similar. We can still address more post 9.0. Ok? Bug fixes only, no optimisations?

File: Slim/Control/Queries.pm [hunk 1]
```diff
@@ -900,6 +873,11 @@ sub albumsQuery {
 						push @artistIds, @{$contributorHash->{$role}->{'id'}};
 					}
 				}
+				# if not dealing with a work, put the main album artist at the top of the list
+				if (!$work) {
+					unshift @artists, $c->{'contributor.name'} if $c->{'contributor.name'};
```
Comment (michaelherger) [inline Slim/Control/Queries.pm]: Shouldn't this be `contributor*s*.name`?

Comment (darrell-k) [reply Slim/Control/Queries.pm]: Yes! Don't know how I didn't spot that!

**Comment by darrell-k** [comment]
`

> Let's try to slow down and stabilise things. We shouldn't shake existing code too much any more, even if it's a more elegant solution or similar. We can still address more post 9.0. Ok? Bug fixes only, no optimisations?

That latest commit is not as scary as git's default diff makes it look. Here is the annotated output of `git diff --minimal -w 189e499441d5cc003e89f348ed6ce8dda6d804f9 1e70684264b12dcae311f379600a7ba4a9f0cb53`:

Does this answer your concerns?
[gitdiff.txt](https://github.com/user-attachments/files/17620262/gitdiff.txt)

File: Slim/Control/Queries.pm (hunk 1 above)
Comment (michaelherger) [reply Slim/Control/Queries.pm]: Do you think there's more left than this? I could merge and fix this line on my end.

Comment (darrell-k) [reply Slim/Control/Queries.pm]: Pushed now, with a couple of other changes as noted in my diff document.

**Comment by michaelherger** [comment]
Let's get that in - thanks!

**Pull request merged**